pkg/controller/controller.go

```diff
@@ -0,0 +1,272 @@
+			},
+			WatchFunc: func(options meta_v1.ListOptions) (watch.Interface, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Service{},
+		0,
+		cache.Indexers{},
+	)
+	remoteCluster := &RemoteCluster{
+		stopCh:          make(chan struct{}),
+		ClusterID:       clusterID,
+		serviceInformer: serviceInformer,
+		queue:           workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+	c.processingMutex.Lock()
+	c.remoteClusters[clusterID] = remoteCluster
+	c.processingMutex.Unlock()
+	serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+	})
+
+	go c.Run(remoteCluster)
+}
+
+func (c *LightHouseController) removeServiceWatcher(clusterID string) {
+	remoteCluster, ok := c.remoteClusters[clusterID]
+	if ok {
+		klog.V(2).Infof("stopping watcher for %s", clusterID)
+		remoteCluster.close()
+		delete(c.remoteClusters, clusterID)
+	}
+}
+
+func (rc *RemoteCluster) close() {
+	close(rc.stopCh)
+	rc.queue.ShutDown()
+	klog.Infof(" watcher queue shutdown for cluster: %s", rc.ClusterID)
+}
+
+func (c *LightHouseController) Run(remoteCluster *RemoteCluster) {
+	defer utilruntime.HandleCrash()
+
+	go remoteCluster.serviceInformer.Run(remoteCluster.stopCh)
+
+	if !cache.WaitForCacheSync(remoteCluster.stopCh, remoteCluster.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	go c.runWorker(remoteCluster)
+}
+
+// HasSynced is required for the cache.Controller interface.
+func (c *RemoteCluster) HasSynced() bool {
+	return c.serviceInformer.HasSynced()
+}
+
+func (c *LightHouseController) runWorker(remoteCluster *RemoteCluster) {
+	for {
+		key, shutdown := remoteCluster.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for cluster \"%s\" stopped", remoteCluster.ClusterID)
+			return
+		}
+		defer remoteCluster.queue.Done(key)
+
+		obj, exists, err := remoteCluster.serviceInformer.GetIndexer().GetByKey(key.(string))
+
+		if err != nil {
+			klog.Errorf("Error processing %s (will retry): %v", key, err)
+			// requeue the item to work on later
+			remoteCluster.queue.AddRateLimited(key)
+		}
+		if !exists {
+			c.ObjectDeleted(obj, remoteCluster.ClusterID)
+			remoteCluster.queue.Forget(key)
+		} else {
+
+			c.ObjectCreated(obj, remoteCluster.ClusterID)
+			remoteCluster.queue.Forget(key)
+		}
+	}
+}
+
+func (c *LightHouseController) ObjectCreated(obj interface{}, clusterID string) {
+	klog.Info("ServiceHandler.ObjectCreated")
+	service := obj.(*core_v1.Service)
+	serviceName := service.ObjectMeta.Name
+	if rs, exists := c.remoteServiceMap.remoteServices[serviceName]; !exists {
```

Comment:
This function is called by a per-cluster worker so `remoteServices` is being accessed by multiple threads and thus needs to be synchronized. However this can be avoided by having a `remoteServices` per `RemoteCluster`. In fact you could even make `remoteServices` local to the `runWorker` function instead of storing as a struct field.

Reply:
Also shouldn't `remoteServices` be keyed by service `Name` and `Namespace`?

Reply:
Some more thoughts ... 

If the federate API had a `CreateOrUpdate` method that gets the existing resource and allows the caller to merge its changes then that would alleviate the need to maintain the `remoteServices` map here. We can consider that later.

The way this code is structured now, a controller restart would basically erase and recreate each `MultiClusterService` piecemeal as it receives each existing service.  The benefit to this approach is that it inherently handles services that were deleted while the controller was down (partially anyway). The potential downside is that it produces unnecessary churn with both federating and for clients observing updates. Not sure if this will be a problem but it's something to keep in mind.  The `CreateOrUpdate` method described above would alleviate it although we'd still have to handle services that no longer exist on restart.